
<!DOCTYPE html>

<html class="writer-html5" data-content_root="../" lang="en">
<head>
<meta charset="utf-8"/><meta content="width=device-width, initial-scale=1" name="viewport"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<title>Quick start: atlasapprox in Python — Explore Average Expression  documentation</title>
<link href="../_static/pygments.css?v=80d5e7a1" rel="stylesheet" type="text/css"/>
<link href="../_static/css/theme.css?v=e59714d7" rel="stylesheet" type="text/css"/>
<link href="../_static/jupyter-sphinx.css" rel="stylesheet" type="text/css"/>
<link href="../_static/thebelab.css" rel="stylesheet" type="text/css"/>
<link href="../_static/sphinx-codeautolink.css?v=125d5c1c" rel="stylesheet" type="text/css"/>
<script src="../_static/jquery.js?v=5d32c60e"></script>
<script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
<script src="../_static/documentation_options.js?v=5929fcd5"></script>
<script src="../_static/doctools.js?v=9bcbadda"></script>
<script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
<script src="../_static/thebelab-helper.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
<script src="../_static/js/theme.js"></script>
<link href="../genindex.html" rel="index" title="Index"/>
<link href="../search.html" rel="search" title="Search"/>
<link href="average.html" rel="next" title="Exploring average gene expression"/>
<link href="../index.html" rel="prev" title="Welcome to atlasapprox in Python"/>
</head>
<body class="wy-body-for-nav">
<div class="wy-grid-for-nav">
<nav class="wy-nav-side" data-toggle="wy-nav-shift">
<div class="wy-side-scroll">
<div class="wy-side-nav-search">
<a class="icon icon-home" href="../index.html">
            Explore Average Expression
          </a>
<div role="search">
<form action="../search.html" class="wy-form" id="rtd-search-form" method="get">
<input aria-label="Search docs" name="q" placeholder="Search docs" type="text"/>
<input name="check_keywords" type="hidden" value="yes"/>
<input name="area" type="hidden" value="default"/>
</form>
</div>
</div><div aria-label="Navigation menu" class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation">
<p class="caption" role="heading"><span class="caption-text">Tutorials:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#"><strong>Quick start: atlasapprox in Python</strong></a><ul>
<li class="toctree-l2"><a class="reference internal" href="#create-virtual-environment"><strong>Create virtual environment</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="#installation"><strong>Installation</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="#initialize-the-api"><strong>Initialize the API</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="#getting-average-gene-expression"><strong>Getting average gene expression</strong></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#output"><strong>Output</strong></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#start-from-scratch"><strong>Start from scratch</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="#more-tutorials-coming-soon"><strong>More tutorials (coming soon)</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="#conclusion"><strong>Conclusion</strong></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="average.html">Exploring average gene expression</a></li>
</ul>
</div>
</div>
</nav>
<section class="wy-nav-content-wrap" data-toggle="wy-nav-shift"><nav aria-label="Mobile navigation menu" class="wy-nav-top">
<i class="fa fa-bars" data-toggle="wy-nav-top"></i>
<a href="../index.html">Explore Average Expression</a>
</nav>
<div class="wy-nav-content">
<div class="rst-content">
<div aria-label="Page navigation" role="navigation">
<ul class="wy-breadcrumbs">
<li><a aria-label="Home" class="icon icon-home" href="../index.html"></a></li>
<li class="breadcrumb-item active"><strong>Quick start: atlasapprox in Python</strong></li>
<li class="wy-breadcrumbs-aside">
<a href="../_sources/pages/quick_start.md.txt" rel="nofollow"> View page source</a>
</li>
</ul>
<hr/>
</div>
<div class="document" itemscope="itemscope" itemtype="http://schema.org/Article" role="main">
<div itemprop="articleBody">
<section id="quick-start-atlasapprox-in-python">
<h1><strong>Quick start: atlasapprox in Python</strong><a class="headerlink" href="#quick-start-atlasapprox-in-python" title="Link to this heading"></a></h1>
<p>The <a class="reference external" href="https://atlasapprox.readthedocs.io/en/latest/index.html"><em>atlasapprox</em></a> API provides access to approximated single-cell data across 28 species, including both animals and plants. You can explore data from species such as <em>Homo sapiens</em> (humans), <em>Mus musculus</em> (mice), <em>Arabidopsis thaliana</em> (thale cress), and <em>Zea mays</em> (corn). Follow this guide to get started with installation, basic usage, and example queries.</p>
<section id="create-virtual-environment">
<h2><strong>Create virtual environment</strong><a class="headerlink" href="#create-virtual-environment" title="Link to this heading"></a></h2>
<p>To ensure consistent dependencies, setting up a virtual environment is recommended before installing the package. Here’s one way to do it:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a virtual environment</span>
<span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">venv</span> <span class="n">myenv</span>

<span class="c1"># Activate your environment (use the appropriate command for your OS)</span>

<span class="c1"># For macOS/Linux users</span>
<span class="n">source</span> <span class="n">myenv</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">activate</span>

<span class="c1"># For Windows users</span>
<span class="n">myenv</span>\<span class="n">Scripts</span>\<span class="n">activate</span> 
</pre></div>
</div>
</section>
<section id="installation">
<h2><strong>Installation</strong><a class="headerlink" href="#installation" title="Link to this heading"></a></h2>
<p>Use <em>pip</em> to install the <em>atlasapprox</em> Python package:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">atlasapprox</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: atlasapprox in /Users/xuxinyi/Desktop/cjenv/lib/python3.10/site-packages (0.2.4)
Requirement already satisfied: requests in /Users/xuxinyi/Desktop/cjenv/lib/python3.10/site-packages (from atlasapprox) (2.32.3)
Requirement already satisfied: numpy in /Users/xuxinyi/Desktop/cjenv/lib/python3.10/site-packages (from atlasapprox) (1.23.0)
</pre></div>
</div>
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: pandas in /Users/xuxinyi/Desktop/cjenv/lib/python3.10/site-packages (from atlasapprox) (2.2.3)
Requirement already satisfied: python-dateutil&gt;=2.8.2 in /Users/xuxinyi/Desktop/cjenv/lib/python3.10/site-packages (from pandas-&gt;atlasapprox) (2.9.0.post0)
Requirement already satisfied: pytz&gt;=2020.1 in /Users/xuxinyi/Desktop/cjenv/lib/python3.10/site-packages (from pandas-&gt;atlasapprox) (2024.2)
Requirement already satisfied: tzdata&gt;=2022.7 in /Users/xuxinyi/Desktop/cjenv/lib/python3.10/site-packages (from pandas-&gt;atlasapprox) (2024.2)
Requirement already satisfied: charset-normalizer&lt;4,&gt;=2 in /Users/xuxinyi/Desktop/cjenv/lib/python3.10/site-packages (from requests-&gt;atlasapprox) (3.4.0)
Requirement already satisfied: idna&lt;4,&gt;=2.5 in /Users/xuxinyi/Desktop/cjenv/lib/python3.10/site-packages (from requests-&gt;atlasapprox) (3.10)
Requirement already satisfied: urllib3&lt;3,&gt;=1.21.1 in /Users/xuxinyi/Desktop/cjenv/lib/python3.10/site-packages (from requests-&gt;atlasapprox) (2.2.3)
Requirement already satisfied: certifi&gt;=2017.4.17 in /Users/xuxinyi/Desktop/cjenv/lib/python3.10/site-packages (from requests-&gt;atlasapprox) (2024.8.30)
Requirement already satisfied: six&gt;=1.5 in /Users/xuxinyi/Desktop/cjenv/lib/python3.10/site-packages (from python-dateutil&gt;=2.8.2-&gt;pandas-&gt;atlasapprox) (1.16.0)
</pre></div>
</div>
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>Note: you may need to restart the kernel to use updated packages.
</pre></div>
</div>
</div>
</div>
</section>
<section id="initialize-the-api">
<h2><strong>Initialize the API</strong><a class="headerlink" href="#initialize-the-api" title="Link to this heading"></a></h2>
<p>Import the <em>atlasapprox</em> Python package and create an API object:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">atlasapprox</span>

<span class="n">api</span> <span class="o">=</span> <span class="n">atlasapprox</span><span class="o">.</span><span class="n">API</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>Data sources for the approximations:

Animals:
- Homo sapiens (human):
  - RNA: Tabula Sapiens (https://www.science.org/doi/10.1126/science.abl4896)
  - ATAC: Zhang et al. 2021 (https://doi.org/10.1016/j.cell.2021.10.024)
- Mus musculus (mouse): Tabula Muris Senis (https://www.nature.com/articles/s41586-020-2496-1)
- Microcebus murinus (mouse lemur): Tabula Microcebus (https://www.biorxiv.org/content/10.1101/2021.12.12.469460v2)
- Caenorhabditis elegans: Cao et al. 2017 (https://www.science.org/doi/10.1126/science.aam8940)
- Crassostrea gigas: Piovani et al. 2023 (https://doi.org/10.1126/sciadv.adg6034)
- Danio rerio (zebrafish): Wagner et al. 2018 (https://www.science.org/doi/10.1126/science.aar4362)
- Clytia hemisphaerica: Chari et al. 2021 (https://www.science.org/doi/10.1126/sciadv.abh1683#sec-4)
- Drosophila melanogaster (fruitfly): Li et al. 2022 (https://doi.org/10.1126/science.abk2432
- Hofstenia miamia: Hulett et al. 2023 (https://www.nature.com/articles/s41467-023-38016-4)
- Isodiametra pulchra: Duruz et al. 2020 (https://academic.oup.com/mbe/article/38/5/1888/6045962)
- Mnemiopsis leidyi: Sebé-Pedrós et al 2018 (https://www.nature.com/articles/s41559-018-0575-6)
- Nematostella vectensis: Steger et al 2022 (https://doi.org/10.1016/j.celrep.2022.111370)
- Prostheceraeus crozieri: Piovani et al. 2023 (https://doi.org/10.1126/sciadv.adg6034)
- Platynereis dumerilii: Achim et al 2017 (https://academic.oup.com/mbe/article/35/5/1047/4823215)
- Strongylocentrotus purpuratus (sea urchin): Paganos et al. 2021 (https://doi.org/10.7554/eLife.70416)
- Spongilla lacustris: Musser et al. 2021 (https://www.science.org/doi/10.1126/science.abj2949)
- Schistosoma mansoni: Li et al. 2021 (https://www.nature.com/articles/s41467-020-20794-w)
- Schmidtea mediterranea: Plass et al. 2018 (https://doi.org/10.1126/science.aaq1723)
- Stylophora pistillata: Levi et al. 2021 (https://www.sciencedirect.com/science/article/pii/S0092867421004402)
- Trichoplax adhaerens: Sebé-Pedrós et al 2018 (https://www.nature.com/articles/s41559-018-0575-6)
- Xenopus laevis: Liao et al. 2022 (https://www.nature.com/articles/s41467-022-31949-2)

Plants:
- Arabidopsis thaliana: Shahan et al 2022 (https://www.sciencedirect.com/science/article/pii/S1534580722000338), Xu et al. 2024 (https://www.biorxiv.org/content/10.1101/2024.03.04.583414v1)
- Lemna minuta: Abramson et al. 2022 (https://doi.org/10.1093/plphys/kiab564)
- Fragaria vesca: Bai et al. 2022 (https://doi.org/10.1093/hr/uhab055)
- Oryza sativa: Zhang et al. 2022 (https://doi.org/10.1038/s41467-021-22352-4)
- Triticum aestivum (wheat): Zhang et al 2023 (https://genomebiology.biomedcentral.com/articles/10.1186/s13059-023-02908-x)
- Zea mays: Marand et al. 2021 (https://doi.org/10.1016/j.cell.2021.04.014)

To hide this message, set the environment variable ATLASAPPROX_HIDECREDITS to any
nonzero value, e.g.:

import os
os.environ["ATLASAPPROX_HIDECREDITS"] = "yes"
import atlasapprox

To propose a new atlas be added to the list of approximations, please contact
Fabio Zanini (fabio DOT zanini AT unsw DOT edu DOT au).
</pre></div>
</div>
</div>
</div>
</section>
<section id="getting-average-gene-expression">
<h2><strong>Getting average gene expression</strong><a class="headerlink" href="#getting-average-gene-expression" title="Link to this heading"></a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">average</span></code> function allows you to retrieve the average gene expression data for selected genes within an organism’s specific organ.</p>
<p>The following example shows how to examine the average expression of five genes (<em>COL13A1</em>, <em>COL14A1</em>, <em>TGFBI</em>, <em>PDGFRA</em>, <em>GZMA</em>) in the human lung:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Call API function with params</span>
<span class="n">avg_gene_expr_lung</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">average</span><span class="p">(</span>
    <span class="n">organism</span> <span class="o">=</span> <span class="s2">"h_sapiens"</span><span class="p">,</span> 
    <span class="n">organ</span> <span class="o">=</span> <span class="s2">"lung"</span><span class="p">,</span> 
    <span class="n">features</span> <span class="o">=</span> <span class="p">[</span><span class="s2">"COL13A1"</span><span class="p">,</span> <span class="s2">"COL14A1"</span><span class="p">,</span> <span class="s2">"TGFBI"</span><span class="p">,</span> <span class="s2">"PDGFRA"</span><span class="p">,</span> <span class="s2">"GZMA"</span><span class="p">],</span> 
    <span class="n">measurement_type</span> <span class="o">=</span> <span class="s2">"gene_expression"</span>
<span class="p">)</span>

<span class="c1"># Display the result</span>
<span class="n">avg_gene_expr_lung</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>neutrophil</th>
<th>basophil</th>
<th>monocyte</th>
<th>macrophage</th>
<th>dendritic</th>
<th>B</th>
<th>plasma</th>
<th>T</th>
<th>NK</th>
<th>plasmacytoid</th>
<th>...</th>
<th>capillary</th>
<th>CAP2</th>
<th>lymphatic</th>
<th>fibroblast</th>
<th>alveolar fibroblast</th>
<th>smooth muscle</th>
<th>vascular smooth muscle</th>
<th>pericyte</th>
<th>mesothelial</th>
<th>ionocyte</th>
</tr>
</thead>
<tbody>
<tr>
<th>COL13A1</th>
<td>0.000000</td>
<td>0.222863</td>
<td>0.000000</td>
<td>0.000711</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.002205</td>
<td>0.000000</td>
<td>0.029147</td>
<td>0.000000</td>
<td>...</td>
<td>0.003937</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.005113</td>
<td>0.446961</td>
<td>0.000000</td>
<td>0.131642</td>
<td>0.067960</td>
<td>0.000000</td>
<td>0.000000</td>
</tr>
<tr>
<th>COL14A1</th>
<td>0.000000</td>
<td>0.000000</td>
<td>0.001422</td>
<td>0.001362</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.002607</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>...</td>
<td>0.007525</td>
<td>0.026666</td>
<td>0.059648</td>
<td>1.110076</td>
<td>1.226022</td>
<td>1.033389</td>
<td>2.108460</td>
<td>0.033580</td>
<td>0.000000</td>
<td>0.000000</td>
</tr>
<tr>
<th>TGFBI</th>
<td>0.065150</td>
<td>0.111107</td>
<td>1.802062</td>
<td>1.252701</td>
<td>2.190132</td>
<td>0.000000</td>
<td>0.083882</td>
<td>0.100460</td>
<td>0.326610</td>
<td>4.492828</td>
<td>...</td>
<td>0.045932</td>
<td>0.067610</td>
<td>0.521915</td>
<td>0.393191</td>
<td>0.175393</td>
<td>0.311884</td>
<td>0.258512</td>
<td>0.119010</td>
<td>0.404976</td>
<td>0.032419</td>
</tr>
<tr>
<th>PDGFRA</th>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000965</td>
<td>0.002414</td>
<td>0.003172</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.005035</td>
<td>0.000000</td>
<td>0.000000</td>
<td>...</td>
<td>0.011427</td>
<td>0.002920</td>
<td>0.000000</td>
<td>1.772957</td>
<td>3.724075</td>
<td>0.128634</td>
<td>0.059852</td>
<td>0.000000</td>
<td>0.332479</td>
<td>0.000000</td>
</tr>
<tr>
<th>GZMA</th>
<td>0.013437</td>
<td>0.142837</td>
<td>0.174047</td>
<td>0.029326</td>
<td>0.020453</td>
<td>0.025113</td>
<td>0.063292</td>
<td>9.006065</td>
<td>19.687157</td>
<td>0.000000</td>
<td>...</td>
<td>0.044351</td>
<td>0.042996</td>
<td>0.073877</td>
<td>0.029919</td>
<td>0.081036</td>
<td>0.119041</td>
<td>0.000000</td>
<td>0.460141</td>
<td>0.044982</td>
<td>0.058806</td>
</tr>
</tbody>
</table>
<p>5 rows × 30 columns</p>
</div></div></div>
</div>
<section id="output">
<h3><strong>Output</strong><a class="headerlink" href="#output" title="Link to this heading"></a></h3>
<p>The function returns a <em>Pandas DataFrame</em> where:</p>
<ul class="simple">
<li><p>Each row represents a gene.</p></li>
<li><p>Each column corresponds to a cell type.</p></li>
<li><p>The values indicate the average gene expression (measured in counts per ten thousand, or cptt).</p></li>
</ul>
</section>
</section>
<section id="start-from-scratch">
<h2><strong>Start from scratch</strong><a class="headerlink" href="#start-from-scratch" title="Link to this heading"></a></h2>
<p>If you’re starting from scratch, the following steps will help you explore the API.</p>
<ol class="arabic simple">
<li><p>Ask about available organisms:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">organisms</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">organisms</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">organisms</span><span class="p">)</span>
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li><p>Ask about available organs within your organism of interest:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">organs</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">organs</span><span class="p">(</span><span class="n">organism</span><span class="o">=</span><span class="s2">"h_sapiens"</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">organs</span><span class="p">)</span>
</pre></div>
</div>
<ol class="arabic simple" start="3">
<li><p>Ask about available cell types within your organism and organ of interest:</p></li>
</ol>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_types</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">celltypes</span><span class="p">(</span><span class="n">organism</span><span class="o">=</span><span class="s2">"h_sapiens"</span><span class="p">,</span> <span class="n">organ</span><span class="o">=</span><span class="s2">"lung"</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cell_types</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>['neutrophil', 'basophil', 'monocyte', 'macrophage', 'dendritic', 'B', 'plasma', 'T', 'NK', 'plasmacytoid', 'goblet', 'AT1', 'AT2', 'club', 'ciliated', 'basal', 'serous', 'mucous', 'arterial', 'venous', 'capillary', 'CAP2', 'lymphatic', 'fibroblast', 'alveolar fibroblast', 'smooth muscle', 'vascular smooth muscle', 'pericyte', 'mesothelial', 'ionocyte']
</pre></div>
</div>
</div>
</div>
</section>
<section id="more-tutorials-coming-soon">
<h2><strong>More tutorials (coming soon)</strong><a class="headerlink" href="#more-tutorials-coming-soon" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>average gene expression</p></li>
<li><p>cell type abundance</p></li>
<li><p>highest measurements by cell type across an organism</p></li>
<li><p>marker genes</p></li>
<li><p>gene coexpression</p></li>
<li><p>gene homologs</p></li>
<li><p>neighborhoods</p></li>
<li><p>gene interaction partners</p></li>
<li><p>…</p></li>
</ul>
</section>
<section id="conclusion">
<h2><strong>Conclusion</strong><a class="headerlink" href="#conclusion" title="Link to this heading"></a></h2>
<p>This tutorial provided a quick start guide to use the <em>atlasapprox</em> Python package. For more detailed information, refer to the official <a class="reference external" href="https://atlasapprox.readthedocs.io/en/latest/python/index.html">documentation</a>.</p>
</section>
</section>
</div>
</div>
<footer><div aria-label="Footer" class="rst-footer-buttons" role="navigation">
<a accesskey="p" class="btn btn-neutral float-left" href="../index.html" rel="prev" title="Welcome to atlasapprox in Python"><span aria-hidden="true" class="fa fa-arrow-circle-left"></span> Previous</a>
<a accesskey="n" class="btn btn-neutral float-right" href="average.html" rel="next" title="Exploring average gene expression">Next <span aria-hidden="true" class="fa fa-arrow-circle-right"></span></a>
</div>
<hr/>
<div role="contentinfo">
<p>© Copyright 2024, Amber Xu.</p>
</div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
</div>
</div>
</section>
</div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
</body>
</html>